File: Platform/Workspace/CSharp/Adapters/Allors.Workspace.Adapters/Core/Workspace/WorkspaceObject.cs
// <copyright file="WorkspaceObject.cs" company="Allors bvba">
// Copyright (c) Allors bvba. All rights reserved.
// Licensed under the LGPL license. See LICENSE file in the project root for full license information.
// </copyright>

namespace Allors.Workspace
{
    using System;
    using Allors.Protocol.Remote.Sync;
    using Allors.Workspace.Meta;
    using System.Linq;
    using Domain;
    using Protocol;
    using Protocol.Remote;

    public class WorkspaceObject : IWorkspaceObject
    {
        private AccessControl[] accessControls;

        private Permission[] deniedPermissions;
        private string sortedAccessControlIds;
        private string sortedDeniedPermissionIds;

        internal WorkspaceObject(Workspace workspace, long objectId, IClass @class)
        {
            this.Workspace = workspace;
            this.Id = objectId;
            this.Class = @class;
            this.Version = 0;
        }

        internal WorkspaceObject(ResponseContext ctx, SyncResponseObject syncResponseObject)
        {
            this.Workspace = ctx.Workspace;
            this.Id = long.Parse(syncResponseObject.i);
            this.Class = (IClass)this.Workspace.ObjectFactory.MetaPopulation.Find(Guid.Parse(syncResponseObject.t));
            this.Version = !string.IsNullOrEmpty(syncResponseObject.v) ? long.Parse(syncResponseObject.v) : 0;
            this.Roles = syncResponseObject.r?.Select(v => new WorkspaceRole(this.Workspace.ObjectFactory.MetaPopulation, v)).Cast<IWorkspaceRole>().ToArray();
            this.SortedAccessControlIds = ctx.ReadSortedAccessControlIds(syncResponseObject.a);
            this.SortedDeniedPermissionIds = ctx.ReadSortedDeniedPermissionIds(syncResponseObject.d);
        }

        public IClass Class { get; }

        public long Id { get; }

        public IWorkspaceRole[] Roles { get; }

        public string SortedAccessControlIds
        {
            get => this.sortedAccessControlIds;
            set
            {
                this.sortedAccessControlIds = value;
                this.accessControls = null;
            }
        }

        public string SortedDeniedPermissionIds
        {
            get => this.sortedDeniedPermissionIds;
            set
            {
                this.sortedDeniedPermissionIds = value;
                this.accessControls = null;
            }
        }

        public long Version { get; private set; }

        IWorkspace IWorkspaceObject.Workspace => this.Workspace;

        public Workspace Workspace { get; }

        public bool IsPermitted(Permission permission)
        {
            if (permission == null)
            {
                return false;
            }

            if (this.accessControls == null && this.SortedAccessControlIds != null)
            {
                this.accessControls = this.SortedAccessControlIds.Split(Encoding.Separator).Select(v => this.Workspace.AccessControlById[long.Parse(v)]).ToArray();
                if (this.deniedPermissions != null)
                {
                    this.deniedPermissions = this.SortedDeniedPermissionIds.Split(Encoding.Separator).Select(v => this.Workspace.PermissionById[long.Parse(v)]).ToArray();
                }
            }

            if (this.deniedPermissions != null && this.deniedPermissions.Contains(permission))
            {
                return false;
            }

            if (this.accessControls != null && this.accessControls.Length > 0)
            {
                return this.accessControls.Any(v => v.PermissionIds.Any(w => w == permission.Id));
            }

            return false;
        }
    }
}

File: Core/Workspace/Typescript.legacy/Site/wwwroot/Core/Workspace/WorkspaceObject.ts
namespace Allors {
  import MetaObject = Meta.MetaObject;
  import ObjectType = Meta.ObjectType;
  import RoleType = Meta.RoleType;

  import SyncResponseObject = Protocol.SyncResponseObject;
  import Compressor = Protocol.Compressor;
  import deserialize = Protocol.deserialize;

  export interface IWorkspaceObject {
        workspace: IWorkspace;
        objectType: ObjectType;
        id: string;
        version: string;
        roleByRoleTypeId: Map<string, any>;

        isPermitted(permission: Permission): boolean;
    }

    export class WorkspaceObject implements IWorkspaceObject {
        workspace: Workspace;
        objectType: ObjectType;
        id: string;
        version: string;
        sortedAccessControlIds: string;
        sortedDeniedPermissionIds: string;

        roleByRoleTypeId: Map<string, any>;

        private cachedSortedAccessControlIds: string;
        private cachedSortedDeniedPermissionIds: string;
        private cachedAccessControls: AccessControl[];
        private cachedDeniedPermissions: Set<Permission>;

        constructor(workspace: Workspace) {
            this.workspace = workspace;
            this.cachedAccessControls = null;
            this.cachedDeniedPermissions = null;
            this.roleByRoleTypeId = new Map();
        }

        new(id: string, objectType: ObjectType) {
            this.objectType = objectType;
            this.id = id;
            this.version = '0';
        }

        sync(
            syncResponseObject: SyncResponseObject,
            sortedAccessControlIdsDecompress: (compressed: string) => string,
            sortedDeniedPermissionIdsDecompress: (compressed: string) => string,
            metaDecompress: (compressed: string) => MetaObject) {

            this.id = syncResponseObject.i;
            this.version = syncResponseObject.v;
            this.objectType = metaDecompress(syncResponseObject.t) as ObjectType;

            this.roleByRoleTypeId = new Map();
            if (syncResponseObject.r) {
                syncResponseObject.r.forEach((role) => {
                    const roleTypeId = role.t;
                    const roleType = metaDecompress(roleTypeId) as RoleType;

                    let value: any = role.v;
                    if (roleType.objectType.isUnit) {
                        value = deserialize(value, roleType.objectType);
                    } else {
                        if (roleType.isMany) {
                            value = (value as string).split(Compressor.itemSeparator);
                        }
                    }

                    this.roleByRoleTypeId.set(roleType.id, value);
                });
            }

            this.sortedAccessControlIds = sortedAccessControlIdsDecompress(syncResponseObject.a);
            this.sortedDeniedPermissionIds = sortedDeniedPermissionIdsDecompress(syncResponseObject.d);
        }

        isPermitted(permission: Permission): boolean {
            if (permission == null) {
                return false;
            }

            if (this.sortedAccessControlIds !== this.cachedSortedAccessControlIds) {
                this.cachedSortedAccessControlIds = this.sortedAccessControlIds;
                if (this.sortedAccessControlIds) {
                    this.cachedAccessControls = this.sortedAccessControlIds
                        .split(Compressor.itemSeparator)
                        .map(v => this.workspace.accessControlById.get(v));
                } else {
                    this.sortedAccessControlIds = null;
                }
            }

            if (this.sortedDeniedPermissionIds !== this.cachedSortedDeniedPermissionIds) {
                this.cachedSortedDeniedPermissionIds = this.sortedDeniedPermissionIds;
                if (this.sortedDeniedPermissionIds) {
                    this.cachedDeniedPermissions = new Set();
                    this.sortedDeniedPermissionIds
                        .split(Compressor.itemSeparator)
                        .forEach(v => this.cachedDeniedPermissions.add(this.workspace.permissionById.get(v)));
                } else {
                    this.cachedDeniedPermissions = null;
                }
            }

            if (this.cachedDeniedPermissions && this.cachedDeniedPermissions.has(permission)) {
                return false;
            }

            if (this.cachedAccessControls) {
                for (const accessControl of this.cachedAccessControls) {
                    if (accessControl.permissionIds.has(permission.id)) {
                        return true;
                    }
                }
            }

            return false;
        }

        invalidate() {
            this.version = '';
        }

    }

}
